qiskit/pulse/commands/parametric_pulses.py

```diff
@@ -0,0 +1,337 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Parametric pulse commands. These are pulse commands which are described by a known formula and some
+parameters.
+
+If a backend supports parametric pulses, it will have the attribute
+`backend.configuration().parametric_pulses`, which is a list of supported pulse shapes, such as
+`['gaussian', 'gaussian_square', 'drag']`. A Pulse Schedule, using parametric pulses, which is
+assembled for a backend which supports those pulses, will result in a Qobj which is dramatically
+smaller than one which uses SamplePulses.
+
+This module can easily be extended to describe more pulse shapes. The new class should:
+  - have a descriptive name
+  - be a well known and/or well described formula (include the formula in the class docstring)
+  - take some parameters (at least `duration`) and validate them, if necessary
+  - implement a `get_sample_pulse` method which returns a corresponding SamplePulse in the
+    case that it is assembled for a backend which does not support it.
+
+The new pulse must then be registered by the assembler in
+`qiskit/qobj/converters/pulse_instruction.py:ParametricPulseShapes`
+by following the existing pattern:
+    class ParametricPulseShapes(Enum):
+        gaussian = commands.Gaussian
+        ...
+        new_supported_pulse_name = commands.YourPulseCommandClass
+"""
+from abc import abstractmethod
+from typing import Any, Dict, Optional
+import math
+
+import numpy as np
+
+from qiskit.pulse.channels import PulseChannel
+from qiskit.pulse.exceptions import PulseError
+from qiskit.pulse.pulse_lib.discrete import gaussian, gaussian_square, drag, constant
+from qiskit.pulse.pulse_lib import continuous
+
+from .command import Command
+from .sample_pulse import SamplePulse
+from .instruction import Instruction
+
+# pylint: disable=missing-docstring
+
+
+class ParametricPulse(Command):
```

Comment:
Shouldn't a `ParametricPulse` and `SamplePulse` both inherit from a common type `Pulse`? Both should also support a call to `draw`.

Reply:
Not all commands are drawable so I don't think it's a requirement, but I am certainly in support of it. Can we make it a separate PR? This one has gone on for a long time and is already large. I think it would be a great candidate for a `good first issue`, `help wanted` tags for the community to pick up

Reply:
`SamplePulse`s are currently drawable and so are `ParametricPulse`s. I still think these should inherit from a common `Pulse` class, which doesn't necessarily need to be drawable.

Reply:
> SamplePulses are currently drawable and so are ParametricPulses.

What makes you say that `ParametricPulse`s are currently drawable?

Reply:
That ParametricPulses define a conversion to a `SamplePulse`.

Reply:
Ah, you could do `my_parametric_pulse.get_sample_pulse().draw()`, yes.

I can add a `Pulse` class

Reply:
`PulseCommand`? I don't know about something being called just `Pulse` when everything is Pulse

Reply:
Sure.